### PR TITLE
Fix branch_stack calculation in `row_bit_count()`

### LINKED ISSUE #8938: [BUG] row_bit_count produces Illegal memory acces on some data sets
**Describe the bug**
If I try to do a row_bit_count on the attached file [BIT_MAP_INPUT_1.parquet.gz](https://github.com/rapidsai/cudf/files/6926527/BIT_MAP_INPUT_1.parquet.gz) (have to unzip it first), then I get Illegal memory access errors.

**Steps/Code to reproduce bug**
Apply the patch

```diff
diff --git a/cpp/tests/transform/row_bit_count_test.cu b/cpp/tests/transform/row_bit_count_test.cu
index 0081cf0d46..01850b0796 100644
--- a/cpp/tests/transform/row_bit_count_test.cu
+++ b/cpp/tests/transform/row_bit_count_test.cu
@@ -16,6 +16,7 @@
 
 #include <cudf/column/column.hpp>
 #include <cudf/column/column_view.hpp>
+#include <cudf/io/parquet.hpp>
 #include <cudf/transform.hpp>
 #include <cudf/types.hpp>
 #include <cudf_test/base_fixture.hpp>
@@ -192,6 +193,18 @@ TEST_F(RowBitCount, StringsWithNulls)
   CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected, *result);
 }
 
+TEST_F(RowBitCount, MapOfStringToListOfStrings)
+{
+  cudf::io::parquet_reader_options opts =
+      cudf::io::parquet_reader_options::builder(cudf::io::source_info("./BIT_MAP_INPUT_1.parquet"))
+      .build();
+
+  cudf::io::table_with_metadata table_with_meta = cudf::io::read_parquet(opts);
+  auto result = cudf::row_bit_count(*table_with_meta.tbl);
+
+  CUDA_TRY(cudaDeviceSynchronize());
+}
+
 std::pair<std::unique_ptr<column>, std::unique_ptr<column>> build_struct_column()
 {
   std::vector<bool> struct_validity{0, 1, 1, 1, 1, 0};
```
and build it (but don't run).

unzip the attached file and place it in your current working directory.  Then run `./gtests/TRANSFORM_TEST` from the test suite.

**Expected behavior**
No crashes no Illegal memory access errors.

### PR DESCRIPTION
**Pull request by mythrocks**
Fixes #8938.

For input with a number of rows exceeding  `max_block_size`, `row_bit_count()` currently
reaches past the bounds of its shared-memory allocation, causing illegal memory access
errors like in [cudf/issues/8938](https://github.com/rapidsai/cudf/issues/8938).

This commit corrects the calculation of the branch stack's base address, and adds a
test for this case.

### CODE REVIEW

**Comment by nvdbaranec** [comment]
Good fix. Thanks.

File: cpp/tests/transform/row_bit_count_test.cu [hunk 1]
```diff
@@ -192,6 +197,66 @@ TEST_F(RowBitCount, StringsWithNulls)
...
+  // Tests that `row_bit_count()` can handle struct<list<int32_t>> with more
+  // than max_block_size (256) rows.
+  // With a large number of rows, computation spills to multiple thread-blocks,
+  // thus exercising the branch-stack comptutation.
+  // The contents of the input column aren't as pertinent to this test as the
+  // column size. For what it's worth, it looks as follows:
+  //   [ struct({0,1}), struct({2,3}), struct({4,5}), ... ]
+
+  using namespace cudf;
+  auto constexpr num_rows = 256 * 2;  // Exceeding a block size.
```
Comment (nvdbaranec) [inline cpp/tests/transform/row_bit_count_test.cu:223]: Hmm.  Interesting issue here.  This code can't know the per-block size that row_bit_count uses so there's no easily-accessible constant.  Maybe just crank this up to something unreasonable for any kind of block size?  Like 2048 or something.

**Review by nvdbaranec — APPROVED** [review]
Just one small note about potentially choosing a row count that will blow any reasonable block size limit.

File: cpp/tests/transform/row_bit_count_test.cu (hunk 1 above)
Comment (mythrocks) [reply cpp/tests/transform/row_bit_count_test.cu:223]: I started out trying to query the device for the maximum number of threads-per-block, and then got distracted.

> Maybe just crank this up to something unreasonable for any kind of block size?

Agreed. That would be a better bet. We're running a 1D kernel, and as per [the CUDA spec](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#thread-hierarchy), we can't have more than 1024 threads per block. It'd be safest to use 2048 or some such.

Thanks for pointing this out. I'll fix this shortly.

**Review by nvdbaranec — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]

**Comment by mythrocks** [comment]
@gpucibot merge

**Comment by mythrocks** [comment]
Thanks for the reviews, chaps. I've merged this change now.

**Pull request merged**